#23546 - Wrong Image opens on HT Report in Safari

mikesmnx opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
Wrong Image opens on HT Report in Safari

### Fixed Issues
$ https://github.com/Expensify/App/issues/23546
PROPOSAL: https://github.com/Expensify/App/issues/23546#issuecomment-1659243256
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->

### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Join #announce channel.
2. Scroll up and open shared images by users.
3. After scrolling a little up again open some other image.
4. Correct image is displayed.

- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
1. Open #announce channel.
2. Scroll up and open shared images by users.
3. After scrolling a little up again open some other image.
4. Correct image is displayed.
5. Turn off network connection.
6. Repeat steps 3-4.

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Join #announce channel (or any other channel with existing shared images).
2. Scroll up and open shared images by users.
3. After scrolling a little up again open some other image.
4. Correct image is displayed.


- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
        - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/12779176/c6592d8e-2d8a-4686-a7f4-99588a1e8709


</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/12779176/e295d6ce-180d-44ef-91fa-157e83f94ede



</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/12779176/427a7481-c005-414c-9a49-6e82c5a446ad




</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->


https://github.com/Expensify/App/assets/12779176/0b4ddc41-30a5-4f14-b713-3a1f9d418ad7



</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/12779176/7daa8ea9-a56a-4757-97cd-e25b55d2a0f6




</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->


https://github.com/Expensify/App/assets/12779176/60132786-2e78-4a0a-a056-48519aa1ca9d



</details>

mikesmnx committed:
#23546 - Wrong Image opens on HT Report in Safari

nkuoch commented on src/components/Attachments/AttachmentCarousel/index.js:
```diff
@@ -189,7 +190,7 @@ function AttachmentCarousel({report, reportActions, source, onNavigate, setDownl
                             // Scroll only one image at a time no matter how fast the user swipes
                             disableIntervalMomentum
-                            pagingEnabled
+                            pagingEnabled={false}
```
can't you just remove it?

mikesmnx replied:
@nkuoch done

mikesmnx committed:
pagingEnabled removed

mikesmnx commented:
I have read the CLA Document and I hereby sign the CLA

mikesmnx commented:
@parasharrajat During testing across all platforms, I discovered that certain properties, such as "bounces" or "maxToRenderPerBatch," slightly altered the current UX. So, I made only the necessary adjustments to maintain the app's existing state.

nkuoch commented:
<img width="648" alt="image" src="https://github.com/Expensify/App/assets/2463975/a618e448-857f-45e5-bb7b-3b745b759025">

mikesmnx commented:
recheck

nkuoch commented:
All good now. @parasharrajat do you mind doing the reviewer checklist?

mikesmnx commented:
@parasharrajat Checking

mikesmnx commented:
@parasharrajat I've found another approach to fix the issue and am currently verifying it across all platforms. I believe it will be ready by tomorrow

mikesmnx committed:
Updated approach

mikesmnx commented:
@parasharrajat Updated the code and chose to retain pagingEnabled. Without it, we would have to implement paging manually, which would result in varied behaviors across supported platforms. I've tested this on the iPhone 11, 13 Pro Max, 14 Pro Max (emulator), Mac desktop app, iOS native app, and also on an Android 13 device, it's working as expected.

mikesmnx commented:
@parasharrajat I've added **contentInset** and **contentInsetAdjustmentBehavior** to achieve more stable behavior (iOS only). Since we don't have any padding, the top, left, bottom, and right are all set to zero (**contentInset**). The **contentInsetAdjustmentBehavior** is set to 'automatic' to ensure that content fits within the scroll area. 

I've switched from **snapToInterval** to **snapToOffsets** because the latter is more stable with large lists. 

I adjusted the **windowSize** (number of rendered items outside visible area) to 1, which has improved memory usage and the selection of the first item. 

I also removed **maxToRenderPerBatch**, as it tends to be unstable with large lists.

mikesmnx committed:
Merge remote-tracking branch 'upstream/main' into mikesmnx-issue-23546

mikesmnx commented:
@parasharrajat Done

mikesmnx committed:
Fixed lint error

mikesmnx committed:
Prettier fix

nkuoch commented:
done

mikesmnx commented:
Hope this helps. I just followed the first three steps from the [issue](https://github.com/Expensify/App/issues/23546) with the iPhone (latest iOS/Safari).

https://github.com/Expensify/App/assets/12779176/96422b18-165b-4afc-9a31-7377a7d1f7ab

mikesmnx committed:
Merge remote-tracking branch 'upstream/main' into mikesmnx-issue-23546

mikesmnx commented on src/components/Attachments/AttachmentCarousel/index.js:
```diff
@@ -195,16 +196,17 @@ function AttachmentCarousel({report, reportActions, parentReportActions, source,
                             snapToAlignment="start"
-                            snapToInterval={containerWidth}
+                            contentInset={{top: 0, left: 0, bottom: 0, right: 0}}
+                            contentInsetAdjustmentBehavior="automatic"
```
Please check my comment [here](https://github.com/Expensify/App/pull/28441#issuecomment-1761303996)

mikesmnx commented on src/components/Attachments/AttachmentCarousel/index.js:
```diff
@@ -195,16 +196,17 @@ function AttachmentCarousel({report, reportActions, parentReportActions, source,
                             initialScrollIndex={page}
-                            initialNumToRender={3}
-                            windowSize={5}
-                            maxToRenderPerBatch={CONST.MAX_TO_RENDER_PER_BATCH.CAROUSEL}
```
Recently, the 'maxToRenderPerBatch' was set to use a constant, but we have removed it in this PR

mikesmnx committed:
Added comment on contentInset

mikesmnx commented on src/components/Attachments/AttachmentCarousel/index.js:
```diff
@@ -195,16 +196,17 @@ function AttachmentCarousel({report, reportActions, parentReportActions, source,
                             pagingEnabled
                             snapToAlignment="start"
-                            snapToInterval={containerWidth}
+                            contentInset={{top: 0, left: 0, bottom: 0, right: 0}}
```
Done

mikesmnx commented:
@parasharrajat I'm not sure I understood correctly. Should I resolve the warning on Android?

mikesmnx commented:
@parasharrajat Apologies for the delay—I was on holiday. I will update the pull request by Wednesday or Thursday

mikesmnx commented:
@parasharrajat done
